docs: add remote-build how-to (#800)
canonical/rockcraft: docs/how-to/outsource-rock-builds-to-launchpad.rst

.. _outsource-rock-builds-to-launchpad:

Outsource rock builds to Launchpad
==================================

> This guide will show you how to outsource your rock builds to Launchpad's `build
> farm <https://launchpad.net/builders>`_ using the ``remote-build`` command. By
> building remotely, you can concurrently assemble your rocks for any of
> Rockcraft's supported architectures.

Reviewer: ```suggestion
This guide shows how to outsource rock builds to the Launchpad `build
farm <https://launchpad.net/builders>`_ with the ``remote-build`` command. By
building remotely, you can concurrently assemble rocks for all supported
architectures.
```

A few notes:
- To keep it task-focused, the introduction might not need to state the command name – the reader will get to it when they enter the instructions.
- We should avoid decorating our products' titles with things like possessives. Less of a writing concern and more a marketing issue – we should preserve our branding.
- It's a good idea to be cautious of "you" and "your", to avoid offense. And from a technical POV, since we're discussing cloud ops, it might not be befitting to say that the user owns the rocks. This is a tricky problem without obvious answers.
- I'm usually going to advise against "use" and "using". :P

Author: https://github.com/canonical/rockcraft/pull/800/commits/909d31e514297a5d628c3bcff2cf1193d788f04b

.. important::

    The ``remote-build`` command is experimental and therefore subject to
    change.

.. _sign-up-for-a-launchpad-account:

Sign up for a Launchpad account
-------------------------------

> To use Rockcraft's ``remote-build`` command, you will need to register for an
> account on `Launchpad <https://launchpad.net>`_.

Reviewer: ```suggestion
To build remotely, you need a
`Launchpad <https://launchpad.net>`_ account.
```

The other changes are more a matter of taste, but the takeaway here is that we can almost always avoid the future tense when describing the current state of affairs.

Author: https://github.com/canonical/rockcraft/pull/800/commits/909d31e514297a5d628c3bcff2cf1193d788f04b

> If you do not already have an account, you can sign up `here

Reviewer: ```suggestion
If you don't already have an account, you can sign up `here
```

See https://docs.ubuntu.com/styleguide/en/#contractions.

Author: https://github.com/canonical/rockcraft/pull/800/commits/909d31e514297a5d628c3bcff2cf1193d788f04b

<https://login.launchpad.net>`_.

Define desired architectures
----------------------------

Once you :ref:`start your remote build <start-a-remote-build>`, all
architectures defined in the :ref:`platforms` key of your project file will be

> built. Rockcraft currently supports AMD64, ARM64, ARM hard float, IA-32, PowerPC

Reviewer 2: I wonder if we should have a separate document describing Rockcraft's supported architectures? It feels like they are mentioned in slightly different ways in multiple places.

Author: I was thinking the same thing. Maybe we could add something like Charmcraft's [platforms reference](https://canonical-charmcraft.readthedocs-hosted.com/en/stable/reference/platforms/) to Rockcraft?

Reviewer: I agree that it should ultimately be separate, but I'd rather not let it stand in the way of this PR.

@jahn-junior Could you create a new ticket for that task?

Reviewer 2: Definitely agree that this is separate work

64-bit little-endian, RISC-V 64-bit and S390x. Your project file can contain any
subset of these architectures.

For example, if you need to build rocks for AMD64, ARM64 and RISC-V 64-bit

> architectures, your project file will include:

Reviewer: ```suggestion
architectures, your project file would declare:
```

"will", "would", and "should" are all difficult, but I prefer "will" least of all because it's an attempt to guarantee future events.

Author: https://github.com/canonical/rockcraft/pull/800/commits/909d31e514297a5d628c3bcff2cf1193d788f04b

.. code-block:: yaml

    platforms:
      amd64:
      arm64:
      riscv64:

.. _start-a-remote-build:

Start a remote build
--------------------

> To start a remote build on Launchpad, your project must be version-controlled by
> Git. Note that the repository doesn't need to be hosted on Launchpad prior to
> build, as Rockcraft will automatically upload the Git repository in the current
> working directory to Launchpad on your behalf.

Reviewer: ```suggestion
To start a remote build on Launchpad, your project must be version-controlled by
Git. Note that the repository doesn't need to be hosted on Launchpad prior to
build, as Rockcraft will automatically upload the Git repository in the current
working directory to Launchpad on your behalf.
```

I see how this could be a step, however it reads to me more like a requirement that the user must take care of before starting this task. What do you think about placing this earlier in the page?

Author: https://github.com/canonical/rockcraft/pull/800/commits/909d31e514297a5d628c3bcff2cf1193d788f04b

> In the root directory of your project, you can now begin your remote build with:

Reviewer: ```suggestion
In the root directory of your project, begin a remote build with:
```

I usually reserve "can" for optional tasks.

Author: https://github.com/canonical/rockcraft/pull/800/commits/909d31e514297a5d628c3bcff2cf1193d788f04b

.. code-block:: bash

    rockcraft remote-build

Due to build queue lengths varying per architecture, you may want to append the
``--launchpad-timeout=<seconds>`` option to stop monitoring the build locally
after a certain amount of time has elapsed. Once timed out on your local

> machine, the build will continue on Launchpad and can be :ref:`recovered

Reviewer: Usually timeouts are a bad thing, but it sounds like it's expected here? With respect to how this paragraph starts out with a conditional, are timeouts an eventuality?

Author: https://github.com/canonical/rockcraft/pull/800/commits/9b22a4845087e5c36c641fce752dd5376814c975

<recover-interrupted-builds>` at any point.

> At this point, you will be asked to acknowledge that all remote builds are

Reviewer: > At this point

It's not clear to me what observable condition we mean here. Can we be more specific? Something like:

> When [x happens], Rockcraft asks whether all remote builds...

Author: https://github.com/canonical/rockcraft/pull/800/commits/909d31e514297a5d628c3bcff2cf1193d788f04b

publicly available on Launchpad.

> .. terminal::

Reviewer 3: This is super cool!

    :input: rockcraft remote-build
    :user: ubuntu

>     :host: rock-dev

Reviewer: Fancy!

    remote-build is experimental and is subject to change. Use with caution.
    All data sent to remote builders will be publicly available. Are you sure you
    want to continue? [y/N]:

> If you are not logged in or have not yet :ref:`registered for Launchpad

Reviewer: ```suggestion
If you aren't logged in or haven't yet :ref:`registered for Launchpad
```

> <sign-up-for-a-launchpad-account>`, you will now be prompted to do so in your

Reviewer: ```suggestion
<sign-up-for-a-launchpad-account>`, Rockcraft asks you to do so in your
```

Author: https://github.com/canonical/rockcraft/pull/800/commits/909d31e514297a5d628c3bcff2cf1193d788f04b

> browser. If this is your first time initiating a remote build from your current

Reviewer: > browser

I usually prefer "web browser" to disambiguate from other kinds.

Author: https://github.com/canonical/rockcraft/pull/800/commits/909d31e514297a5d628c3bcff2cf1193d788f04b

> machine, you will then be asked to authorize access to your Launchpad account.

Reviewer: > machine

Another tricky term in cloud ops. The user might already be interacting through a remote machine, or it might be a runner that's sending the commands, so I often reach for "host".

Author: https://github.com/canonical/rockcraft/pull/800/commits/909d31e514297a5d628c3bcff2cf1193d788f04b

> Once authorized, your project will be uploaded to Launchpad and placed in the

Reviewer: ```suggestion
Once authorized, your project is uploaded to Launchpad and placed in the
```

Author: https://github.com/canonical/rockcraft/pull/800/commits/909d31e514297a5d628c3bcff2cf1193d788f04b

build queues for each architecture defined in your project file.
Unless a local timeout was specified, the status of each build will be

> continuously monitored and reported back to you.

Reviewer 3: You can also hit ctrl+c to either cancel the build or leave it running and recover it later.

Author: Good point. Should I say "Unless interrupted locally..." to be more general?

Reviewer 3: I'm more thinking that can be added as a paragraph or section, especially since it offers you whether to do cleanup or not.

Author: https://github.com/canonical/rockcraft/pull/800/commits/9b22a4845087e5c36c641fce752dd5376814c975

Let me know what you think about the changes. Open to any feedback on it.

Check the build results
-----------------------

Once all of your builds have either built successfully or failed, your rocks

> will be downloaded to the root of your project along with their build logs.

Reviewer: ```suggestion
are downloaded to the root of your project along with their build logs.
```

Author: https://github.com/canonical/rockcraft/pull/800/commits/909d31e514297a5d628c3bcff2cf1193d788f04b

> Your completed build can also be viewed on Launchpad by navigating to:

Reviewer: > navigating

Consider "going". I think it's a bit more friendly.

Author: https://github.com/canonical/rockcraft/pull/800/commits/909d31e514297a5d628c3bcff2cf1193d788f04b

.. code-block:: text

    https://code.launchpad.net/~<user>/<user>-craft-remote-build/+git/<build-id>

> where ``<user>`` is your Launchpad username and ``<build-id>`` is the ID

Reviewer: ```suggestion
Replace ``<user>`` with your Launchpad username and ``<build-id>`` with the ID
```

Completely optional, just something to consider to make it more action-oriented.

Author: https://github.com/canonical/rockcraft/pull/800/commits/909d31e514297a5d628c3bcff2cf1193d788f04b
